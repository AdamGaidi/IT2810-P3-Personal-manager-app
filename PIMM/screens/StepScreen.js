import Expo from "expo";
import React from "react";
import {
  ScrollView,
  StyleSheet,
  View,
  Text,
  TextInput,
  Image,
  RefreshControl
} from "react-native";
import ProgressBarContainer from "../containers/ProgressBarContainer";
import StatisticsContainer from "../containers/StatisticsContainer";
import { Pedometer } from "expo";
import Colors from "../constants/Colors";
import Strings from "../constants/Strings";

export default class StepScreen extends React.Component {
  state = {
    pedometerAvailable: "checking",
    stepsToday: 0,
    stepsThisWeek: 0,
    dailyGoal: 2131,
    pedometerError: false,
    refreshing: false
  };

  static navigationOptions = {
    title: "Steps"
  };

  componentDidMount() {
    this._subscribe();
  }

  componentWillUnmount() {
    this._unsubscribe();
  }

  updateStepsToday = () => {
    const start = new Date();
    start.setHours(0, 0, 0, 0);
    const end = new Date();
    end.setHours(24, 0, 0, 0);

    Pedometer.getStepCountAsync(start, end).then(
      result => {
        this.setState({ stepsToday: result.steps });
      },
      error => {
        this.setState({ pedometerError: true });
      }
    );
  };

  updateStepsThisWeek = () => {
    const end = new Date();
    const start = new Date();
    start.setDate(end.getDate() - 7);

    Pedometer.getStepCountAsync(start, end).then(
      result => {
        this.setState({ stepsThisWeek: result.steps });
      },
      error => {
        this.setState({ pedometerError: true });
      }
    );
  };

  _subscribe = () => {
    this._subscription = Pedometer.watchStepCount(result => {
      this.updateStepsToday();
      this.updateStepsThisWeek();
    });

    Pedometer.isAvailableAsync().then(
      result => {
        this.setState({
          pedometerAvailable: String(result)
        });
      },
      error => {
        this.setState({
          pedometerError: true
        });
      }
    );
    this.updateStepsToday();
    this.updateStepsThisWeek();
    this.setState({ refreshing: false });
  };

  _unsubscribe = () => {
    this._subscription && this._subscription.remove();
    this._subscription = null;
  };

  updatePedometer = () => {
    this.setState({ refreshing: true, pedometerError: false });
    this._subscribe();
    this.setState({ refreshing: false });
  };

  renderErrorMessage() {
    return (
      <ScrollView
        style={styles.container}
        refreshControl={
          <RefreshControl
            refreshing={this.state.refreshing}
            onRefresh={this.updatePedometer}
          />
        }
      >
        <View style={styles.stepsContainer}>
          <Text style={styles.quoteText}>{Strings.pedometerUnavailable}</Text>
        </View>
      </ScrollView>
    );
  }

  isGoalReached() {
    return this.state.stepsToday >= this.state.dailyGoal;
  }

  returnMotivationQuote() {
    if (this.isGoalReached()) {
      return Strings.goalReached.true;
    } else {
      return Strings.goalReached.false;
    }
  }

  render() {
    if (!this.state.pedometerError) {
      return (
        <ScrollView style={styles.container}>
          <View style={[styles.stepsContainer]}>
            <Text style={styles.quoteText}>{this.returnMotivationQuote()}</Text>
            <ProgressBarContainer
              dailyGoal={this.state.dailyGoal}
              stepsToday={this.state.stepsToday}
            />
            <View>
              <StatisticsContainer
                stepsToday={this.state.stepsToday}
                stepsThisWeek={this.state.stepsThisWeek}
              />
            </View>
            <View style={styles.container}>
              <Image
                style={styles.statisticsImage}
                source={require("../assets/images/walking-man.jpg")}
              />
            </View>
            {
              (motivationText = this.isGoalReached() ? (
                <Text style={styles.quoteText}>
                  {Strings.motivation.goalReached}{" "}
                </Text>
              ) : (
                <Text style={styles.quoteText}>
                  {Strings.motivation.goalNotReached}
                </Text>
              ))
            }
          </View>
        </ScrollView>
      );
    } else {
      return this.renderErrorMessage();
    }
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.appBackground
  },
  stepsContainer: {
    flex: 1,
    paddingVertical: 30,
    paddingHorizontal: 20
  },
  quoteText: {
    textAlign: "center",
    paddingHorizontal: 15,
    paddingBottom: 10,
    color: Colors.primaryBlue,
    fontSize: 18,
    lineHeight: 20
  },
  statisticsImage: {
    width: null,
    resizeMode: "contain",
    height: 220
  }
});

Expo.registerRootComponent(StepScreen);
