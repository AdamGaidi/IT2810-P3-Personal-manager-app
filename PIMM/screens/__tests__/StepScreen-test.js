import React from "react";
import StepScreen from "../StepScreen";
import ShallowRenderer from "react-test-renderer/shallow";

it("renders shallow correctly without props", () => {
  const renderer = new ShallowRenderer();
  renderer.render(<StepScreen/>);
  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
