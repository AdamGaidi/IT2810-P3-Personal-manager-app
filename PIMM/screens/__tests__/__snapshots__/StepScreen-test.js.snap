// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders shallow correctly without props 1`] = `
<ScrollViewMock
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <Component
    style={
      Array [
        Object {
          "flex": 1,
          "paddingHorizontal": 20,
          "paddingVertical": 30,
        },
      ]
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#5a95ff",
          "fontSize": 18,
          "lineHeight": 20,
          "paddingBottom": 10,
          "paddingHorizontal": 15,
          "textAlign": "center",
        }
      }
    >
      Walk a little bit more!
    </Text>
    <ProgressBarContainer
      dailyGoal={10000}
      stepsToday={0}
    />
    <Component>
      <StatisticsContainer
        stepsThisWeek={0}
        stepsToday={0}
      />
    </Component>
    <Component
      style={
        Object {
          "backgroundColor": "#fff",
          "flex": 1,
        }
      }
    >
      <Image
        source={1}
        style={
          Object {
            "height": 220,
            "resizeMode": "contain",
            "width": null,
          }
        }
      />
    </Component>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#5a95ff",
          "fontSize": 18,
          "lineHeight": 20,
          "paddingBottom": 10,
          "paddingHorizontal": 15,
          "textAlign": "center",
        }
      }
    >
      Walking is good for your body! 
Go walk a little more.
    </Text>
  </Component>
</ScrollViewMock>
`;
